### DISCUSSION TITLE
End a task with a listener (Or a way to stop a task with the InventoryCloseEvent ?)

### DISCUSSION THREAD

**Discussion by Traquolix** [Minestom/Minestom#1769] [Help]
I'm trying to make a trade menu.
I want the trade *process* to stop once one of the players close the inventory, and give back the items etc.

Here is the process of the trade :
- Player 1 put items in the trade menu / Player 2 put items in the trade menu
- Player 1 accept the trade
- Player 2 accept the trade
- There is a 3 seconds countdown 
-> If a player close their inventory before the end of that countdown, the trade is cancelled if not, the trade proceeds.
- Trade done / cancelled.

I got the part of closing the inventory before the countdown working, since I'm not using the scheduler so I can just listen to InventoryCloseEvent. But how to do so in a task ? Or maybe, is there any way to make that kind of mechanic without tasks/scheduler ? Because it seems like the scheduler doesn't care about any listeners.

I tried to use SubmitTask(), but the stop conditions doesn't work. When I call a player.getOpenInventory(), it will never return a -1, even if the player doesn't have an inventory opens at the moment. The Event doesn't seem to fire (or get detected by the listener who is running), not until the task is finally complete at least. 

Any idea on how to approach / fix this problem ? Maybe using tasks is the wrong choice, or maybe I'm doing *something* wrong, would not be surprised.

I know the schedulers & tasks are synchronous, which probably doesn't help. So maybe another solution will be needed. I just don't know which one.

Thank you very much in advance.

**Comment by reeve567** [comment]
If you're still working on this, the easiest way to do this seems like it'd be just storing something like a trade cancelled variable,
I haven't gotten around to working with tasks in Minestom yet, but how I'd do it with paper is to cancel the task after storing the task in a variable. I'm not sure what your code structure looks like, but you should be able to pass things into the task scope, so that you can have the listener outside and then store whether it's been used or not.

**Reply by Tofaa2** [reply]
private Task task;

public void handleYourListener() {
  task.cancel();
}
